Dart/src/com/jetbrains/lang/dart/psi/impl/DartFileReference.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2000-2015 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.jetbrains.lang.dart.psi.impl;
+
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiReference;
+import com.intellij.util.IncorrectOperationException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Reference to a file in import or export directive.
+ */
+public class DartFileReference implements PsiReference {
+  @NotNull private final PsiElement myElement;
+  @NotNull private final String myUri;
+  @NotNull private final PsiFile myPsiFile;
+  @NotNull private final TextRange myRange;
+
+  public DartFileReference(@NotNull final DartUriElementBase uriRefExpr, @NotNull final String uri, @NotNull final PsiFile psiFile) {
+    final int offset = uriRefExpr.getText().indexOf(uri);
+    assert offset >= 0 : uriRefExpr.getText() + " doesn't contain " + uri;
+
+    myElement = uriRefExpr;
+    myUri = uri;
+    myPsiFile = psiFile;
+    myRange = TextRange.create(offset, offset + uri.length());
+  }
+
+  @NotNull
+  @Override
+  public PsiElement getElement() {
+    return myElement;
+  }
+
+  @Override
+  public TextRange getRangeInElement() {
+    return myRange;
+  }
+
+  @Nullable
+  @Override
+  public PsiElement resolve() {
+    return myPsiFile;
+  }
+
+  @NotNull
+  @Override
+  public String getCanonicalText() {
+    return myUri;
+  }
+
+  @Override
+  public PsiElement handleElementRename(final String newElementName) throws IncorrectOperationException {
+    throw new IncorrectOperationException("Can't rename file");
```

Comment:
This completely blocks file rename/move functionality. `return myElement` is better - file will be moved/renamed, but uri will become invalid.
Would be great to update uri on file move/rename. With IDE-driven resolution it works perfectly. I wouldn't like to loose the feature.